uap_tracker/video_tracker.py

```diff
@@ -121,54 +138,76 @@ def update_trackers(self, tracker_type, bboxes, frame):
+        tic3 = time.perf_counter()
+        #print(f"{frame_count}: Resizing frame {tic3 - tic2:0.4f} seconds")
 
         frame_gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
+        tic4 = time.perf_counter()
+        #print(f"{frame_count}: Converting to gray scale {tic4 - tic3:0.4f} seconds")
 
-        # Blur image
-        if self.blur:
-            print(f"Applying blue radius:{self.blur_radius}")
-            frame_gray = cv2.GaussianBlur(
-                frame_gray, (self.blur_radius, self.blur_radius), 0)
-            # frame_gray = cv2.medianBlur(frame_gray, self.blur_radius)
+        frame_gray = utils.noise_reduction(self.noise_reduction, frame_gray, self.blur_radius)
+        tic5 = time.perf_counter()
+        #print(f"{frame_count}: Reducing noise {tic5 - tic4:0.4f} seconds")
 
         self.frames = {
             'original': frame,
             'grey': frame_gray
         }
+
         if self.detection_mode == 'background_subtraction':
-            keypoints, frame_masked_background = self.keypoints_from_bg_subtraction(
-                frame_gray)
+
+            tic6 = time.perf_counter()
+            keypoints, frame_masked_background = self.keypoints_from_bg_subtraction(frame_gray)
             if frame_count < 5:
                 # Need 5 frames to get the background subtractor initialised
                 return
             self.frames['masked_background'] = frame_masked_background
             bboxes = [utils.kp_to_bbox(x) for x in keypoints]
+            tic7 = time.perf_counter()
+            #print(f"{frame_count}: Background subtraction {tic7 - tic6:0.4f} seconds")
+
             if self.calculate_optical_flow:
-                self.frames['optical_flow'] = self.optical_flow(frame_gray)
+                # Mike: The optical flow stuff apears to just add the frame to the frames list, so is an ideal candidate
+                # to perform on a different thread
+                optical_flow_thread = Thread(target=self.perform_optical_flow_task, args=(frame_count, frame_gray, tic7))
+                optical_flow_thread.start()
+                worker_threads.append(optical_flow_thread)
         else:
             bboxes = []
             keypoints = []
 
         self.keypoints = keypoints
+        tic9 = time.perf_counter()
 
         self.update_trackers(self.tracker_type, bboxes, frame)
+        tic10 = time.perf_counter()
+        #print(f"{frame_count}: Updating trackers {tic10 - tic9:0.4f} seconds")
 
         frame_count + 1
 
+        # Mike: Wait for worker threads to join before publishing events as their results might be required
+        for worker_thread in worker_threads:
+            worker_thread.join()
+
         if self.events is not None:
             self.events.publish_process_frame(self)
 
+        tic11 = time.perf_counter()
+        #print(f"{frame_count}: Publishing process frame event {tic11 - tic10:0.4f} seconds")
```

Comment:
If you're just trying to time a block of code, a cleaner way is to use a with block:
https://stackoverflow.com/questions/33987060/python-context-manager-that-measures-time

Reply:
Okay, I will update that in a future branch if that is okay. Thanks for the pointer!

Reply:
I have added a TODO for this and will include this change in my next set of changes. I just need to try and get a handle on what its doing and how to use it.